rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/monitoring/views/CompactFlowPanelOwn.java

```diff
@@ -0,0 +1,194 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive_grid_cloud_portal.rm.client.monitoring.views;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import org.ow2.proactive_grid_cloud_portal.rm.client.CompactView;
+import org.ow2.proactive_grid_cloud_portal.rm.client.NodeSource;
+
+
+/**
+ * It is flow panel to draw compact representation of hodesources/hosts/nodes
+ * where all nodes are nodes own by the current user.
+ * In comparing to CompactFlowPanel, this class also treats nodesources as
+ * ephemeral items. So whenever CompactView asks to draw nodesource, it is just stored
+ * in the nodeSourceTiles. And then if at some point we have a node to draw we will
+ * draw nodesource as well.
+ */
+public class CompactFlowPanelOwn extends CompactFlowPanel {
+
+    private Map<String, CompactView.Tile> nodeSourceTiles = new HashMap<>();
+
+    public CompactFlowPanelOwn() {
+        super();
+    }
+
+    @Override
+    public void drawNodeSource(CompactView.Tile nsTile) {
+        nodeSourceTiles.put(nsTile.getNodesource().getSourceName(), nsTile);
+    }
+
+    @Override
+    public void remove(NodeSource nodeSource) {
+        nodeSourceTiles.remove(nodeSource.getSourceName());
+        super.remove(nodeSource);
+    }
+
+    @Override
+    protected void drawDeployingNode(CompactView.Tile nodeTile) {
+        // own nodes are nodes that already been deployed and used by user
+    }
+
+    @Override
+    public void redrawNodeSource(NodeSource nodeSource) {
+        nodeSourceTiles.get(nodeSource.getSourceName()).refresh(nodeSource);
+        if (isNodeSourceDrawn(nodeSource.getSourceName())) {
+            super.redrawNodeSource(nodeSource);
+        }
+    }
+
+    /**
+     * draws a node and if necessary draws nodesource, and if necessary draws host
+     * @param nodeTile
+     * @param hostTile
+     */
+    @Override
+    protected void drawNormalNode(CompactView.Tile nodeTile, CompactView.Tile hostTile) {
+        int index = 0;
+
+        if (!isNodeSourceDrawn(nodeTile.getNode().getSourceName())) {
+            final CompactView.Tile nsTile = nodeSourceTiles.get(nodeTile.getNode().getSourceName());
+            model.add(new HierarchyNodeSource(nsTile.getNodesource()));
+            model.sort(Comparator.comparing(a -> a.getNodeSource().getSourceName()));
+            this.insert(nsTile, indexOf(nsTile.getNodesource()).get());
+        }
+
+        for (HierarchyNodeSource hierarchyNodeSource : model) {
+            if (hierarchyNodeSource.getNodeSource().getSourceName().equals(nodeTile.getNode().getSourceName())) {
```

Comment:
Maybe some parts in this for loop can be extracted to the upper class and used both here and in the `super` implementation?

Reply:
I agree with justine... and at least split this function into sub private ones. Same remark for (*).

Reply:
I do not see how I can do this. But I will try